src/main/java/week3/WordSearch.java

```diff
@@ -0,0 +1,195 @@
+   *    ● You can start at any position.
+   *    ● You can move to one of the 8 adjacent cells (horizontally/vertically/diagonally).
+   *    ● You can't visit the same cell twice in the same word.Assume you have a dictionary class
+   *      with these two methods:
+   *    ● isWord(string): Returns whether the given string is a valid word.
+   *    ● isPrefix(string): Returns whether the given string is a prefix of at least one word in
+   *      the dictionary.
+   *
+   * Your function receives the number of rows, number of columns, a 2-dimensional array of
+   * characters (of the native char data type), and the dictionary. You should return the set of all
+   * words found.
+   */
+
+  private int rows;
+  private int cols;
+
+  /*
+   * Iteratively try to find longer words
+   * initializeMap = get all valid starting points on the grid (valid prefixes of length 1)
+   *
+   * Iteration: while there are longer words found, try to append the next character to the valid
+   * prefixes in the set.
+   *
+   * Time Complexity: Initialization + (number of iterations * cost of each iteration)
+   *                  O(|grid|*Words) + (O(|grid|) * O(|grid|*|grid| * |words|))
+    *                  O(|grid|^3*|words|)
+   *
+   *      where |grid| = rows*cols is the size of the grid and
+   *            |Words| = size of the dictionary / number of words in the dictionary
+   *
+   *  Intuitively, at worst the whole grid forms a word and thus we iterate over the while loop
+   *  |grid| times.
+   *
+   *  In each iteration, we can have all the grid being a prefix of some word, and the processing of
+   *  each path involves O(1) to get adjacent cells and O(|words|) to validate newPrefix
+   */
+  public Set<String> findWords(char[][] grid, Dictionary dictionary) {
+
+    if (grid == null || dictionary == null) {
+      return new HashSet<>();
+    }
+
+    Set<String> words = new HashSet<>();
+    Map<String, List<Path>> prefixes = initializeMap(dictionary, grid);
+
+    while (prefixes.size() > 0) {
+      Map<String, List<Path>> newPrefixes = new HashMap<>();
+
+      prefixes.forEach((p, paths) ->                                //O(|grid|)
+          newPrefixes.putAll(
+              paths
+                  .stream()                                         //O(|grid|)
+                  .flatMap(c -> getUnvisitedAdjCells(c).stream())   //O(1)
+                  .filter(c -> {
+                    String newPrefix = p + grid[c.x][c.y];          // O(1)
+
+                    if (dictionary.isWord(newPrefix)) {             //O(|words|)
+                      words.add(newPrefix);                         //O(1)
+                    }
+
+                    return dictionary.isPrefix(newPrefix);          //O(|words|)
+                  })
+                  .collect(Collectors.groupingBy(c -> p + grid[c.x][c.y]))
+          )
+      );
+
+      prefixes = newPrefixes;
+    }
+
+    return words;
+  }
```

Comment:
I would have accepted a recursive solution, since it's simple and the grids have to get pretty large before recursion depth is a problem (as long as you identify that as a potential problem). Having said that, demonstrating how you'd produce a more scalable solution is great. Thanks.